### PR TITLE
scsynth: macOS: clip values on hw out busses

### PR DESCRIPTION
**Pull request by elgiano**
## Purpose and Motivation

Related to #4894.
Proposes a fix for CoreAudio handling of loud volumes. Since CoreAudio only applies a relative attenuation, setting a low system volume doesn't prevent output values > +- 1 from sounding absolutely loud, possibly leading to hear damage.
This PR attempts to fix it by:
- introducing a new ServerOption + cli switch: `-s safetyClipThreshold`
- clipping values between +-safetyClipThreshold when CoreAudioDriver writes out
- safetyClipThreshold defaults to 2.0 to allow some headroom and save some ears

Although clipping affects inter-app routing (as it happens through the same out busses), it doesn't affect recording.

Note: `s.options.safetyClipThreshold = inf` is valid (thx c++!) and makes clipping ineffective

## Needs discussion: cross-platform
This is initially macos only. Any consideration about making it general? I wouldn't like to have it on linux for instance, where it is not needed, but maybe it would be good to have the same behavior for all platforms?

## Types of changes

- Documentation
- New feature
- Breaking change 
( doesn't change any interface, but introduces hard clipping on macos scsynth by default and adds a cli switch )
## To-do list

- [ ] Code is tested
- [ ] All tests are passing
- [x] Updated documentation
- [ ] This PR is ready for review

### CODE REVIEW

File: SCClassLibrary/Common/Control/Server.sc [hunk 1]
```diff
@@ -222,6 +225,9 @@ ServerOptions {
 		if (maxLogins.notNil, {
 			o = o ++ " -l " ++ maxLogins;
 		});
+		if (safetyClipThreshold.notNil, {
+			o = o ++ " -s " ++ safetyClipThreshold;
```
Comment (dyfer) [inline SCClassLibrary/Common/Control/Server.sc:229]: I think it would be good to have a platform switch here (just for adding the `-s` option)

**Comment by dyfer** [comment]
Thanks @elgiano !

> I wouldn't like to have it on linux for instance, where it is not needed, but maybe it would be good to have the same behavior for all platforms?

IMO it should be macOS only, as other backends seem to clip by themselves (AFAIU).

I'd also like to see a way to disable clipping completely, as in omitting `sc_clip2` to preserve current behavior. We could use `-s 0` or `-s -1` for that, although maybe that's a little counterintuitive. 

I'd also prefer to have less headroom - full scale signal with high RMS will still be much louder than other sounds and I don't think we need 6dB (`2.0`) headroom, but maybe 3 or 2dB. But if others want to stick with 6dB I'm fine with that.

**Comment by elgiano** [comment]
I don't know how to disable clip completely. I mean, I could put an "if", but that would be one more "if" that wasn't there before anyway. Isn't -s inf a fair enough bypass switch? Comparing to inf will never clip

**Comment by dyfer** [comment]
> I don't know how to disable clip completely. I mean, I could put an "if", but that would be one more "if" that wasn't there before anyway. Isn't -s inf a fair enough bypass switch? Comparing to inf will never clip

Yeah, I can't say that I know how to do it either. One ugly, but likely well-performing way would be to put `if` around the whole `for (int s = 0, b = 0; b < numOutputBuses && s < numOutputStreams; s++) {` loop (and make a copy of it without clip) but that would be a lot of code duplication.

**Comment by elgiano** [comment]
Idea: one way could be to create a subclass of SC_CoreAudioDriver, say SC_CoreAudioClippingDriver, override only the Run method, and select which class to use in here: 
https://github.com/supercollider/supercollider/blob/f806ace7bd8565dd174e7d47a1b32aaa4175a46e/server/scsynth/SC_CoreAudio.cpp#L473

**Comment by dyfer** [comment]
> Idea: one way could be to create a subclass of SC_CoreAudioDriver, say SC_CoreAudioClippingDriver, override only the Run method, and select which class to use in here:
> https://github.com/supercollider/supercollider/blob/f806ace7bd8565dd174e7d47a1b32aaa4175a46e/server/scsynth/SC_CoreAudio.cpp#L473

Yeah, that might be a good way to go about it...

**Comment by elgiano** [comment]
I've written a POC of the subclass implementation.
The cost of it is to make SC_CoreAudioDriver::Run a virtual function, so that it can be overwritten by SC_CoreAudioClippingDriver.

I'm not sure if it's worth it, I have a feeling we should avoid having a virtual Run function. I'm not an expert at this level of precision, so any help would be fundamental. Virtuality help avoiding code duplication... but is the vtable bad for Run's performance? Do anyone know any other way around this?

**Comment by jamshark70** [comment]
> IMO it should be macOS only, as other backends seem to clip by themselves (AFAIU).

I don't recall if we tested other OSes rigorously for this. SC+JACK in Linux won't overload because JACK has no volume control -- but I bet you could replicate the problem by running SC --> another app that scales down volume --> hardware.

If I have time, I might also try it with the experimental PulseAudio backend: PA *does* have a volume control.

Did anyone check Windows?

The test would be: Turn system volume down, and try signals with amp > 1. If it gets louder out of the speakers at +10, +20 etc dB, then we knew the system isn't clipping internally.

> I'd also prefer to have less headroom - full scale signal with high RMS will still be much louder than other sounds and I don't think we need 6dB (2.0) headroom, but maybe 3 or 2dB.

I proposed the small headroom -- I wouldn't mind reducing it to 2 dB.

**Comment by elgiano** [comment]
> I don't recall if we tested other OSes rigorously for this. SC+JACK in Linux won't overload because JACK has no volume control -- but I bet you could replicate the problem by running SC --> another app that scales down volume --> hardware.

On Linux it's managed by alsa, right ?(one level deeper than jack). I don't know how it's done internally, but I do know that a blown-up filter sounds quiet when system volume is quiet.

> The test would be: Turn system volume down, and try signals with amp > 1. If it gets louder out of the speakers at +10, +20 etc dB, then we knew the system isn't clipping internally.

Great! I usually tested this with something like what was reported to be harmful on the forum:
```supercollider
{ LPF.ar(WhiteNoise.ar, -200) }.play // this is super loud, it spikes around 1e14
```

**Comment by jamshark70** [comment]
Using the PulseAudio backend from #5081, I can reproduce the same problem that macOS users have reported.

The TL;DR conclusion, then, is that it isn't safe to implement this only for macOS -- we should do it for all platforms.

The issue, as I understand it, is:

- Floating-point signal that is typically +/- 1, but with the possibility of far exceeding that range.
- --> Volume control set to some low level, say -20 or -40 dB (reducing the sound pressure level by a factor of 10 or 100 respectively).
- --> Hardware. (We definitely expect clipping to happen in the hardware -- soundcards usually work with integer sample formats, where it's literally impossible to represent out of range sample values.)

It's not at all far-fetched to imagine a user generating full-scale signals in SC, and using the system volume control to bring that down to a listenable level. User expectation at this point is, "I set my volume control to -40 dB so I shouldn't get anything louder than that in my earphones."

If SC accidentally generates a +700 dB signal (which I've done before), then the volume control makes that +640 dB and the hardware clips at full-scale -- but the user turned down the volume because full-scale is too loud = ear damage.

The defining feature of the issue, then, is that system volume control and pass through a signal louder than the set level.

To test, I ran this on scsynth built for PulseAudio:

```supercollider
(
a = {
	var trig = Impulse.kr(2);
	var amp = Demand.kr(trig, 0, Dseq([Dseries(-18, 6, 8)], inf));
	
	SinOsc.ar(440, 0, Lag.kr(amp, 0.1).dbamp).dup
}.play;
)
```

-18, -12, -6, 0  -- no clipping
+6, +12, +18, +24 -- clipping

Then I went to the PulseAudio volume panel.

When the scsynth volume level is set in pavucontrol to 0 dB, as expected, I hear four clean tones and four distorted tones.

If I turn pavucontrol-scsynth volume down to -20 dB, then I hear *seven* clean tones and one distorted tone -- and the output level goes above -20 dB.

If the volume control is clipping as we want (to protect hearing), then I would hear four clean and four distorted tones, just quieter.

That's definitely not the result I got.

Therefore, the same problem can occur in Linux.

**Comment by jamshark70** [comment]
Moreover, if the intention is ear safety, the end-stage clipper needs to be based on the server's volume setting, not full scale.

Same scenario: user's speakers/headphones are loud. Only now they discover, "if I write `s.volume = -40` in my startup file, then SC is at a good level and I don't have to think about it"... until getting a super loud signal, which SC's volume control reduces slightly and SC's end stage clipping brings to... full scale again, *bang* and "why didn't that volume control protect my hearing?"

I.e. you can imagine this case purely within SC, irrespective of OS.

That makes implementation trickier... or carefully document the limitations of SC's internal volume control (which is tricky to message in a way that doesn't look dumb).

**Comment by dyfer** [comment]
I think it's a valid point about volume setting on the server... but you
could see it both ways - master volume to prevent everything from clipping,
or actual master volume that limits output. I think the former is more
common.
Regardless, this would be a consideration for the sclang-side limiter to
address, independent from this PR, IMO.

**Comment by elgiano** [comment]
So, maybe we shouldn't deal with this at the backed level, and focus on solving the issues with a language-side implementation. After having tried the server path, I'm much more willing to abandon it.
Server pros:
- backends that don't need clipping are unaffected
- backends that need clipping are reliably clipped
- doesn't affect recording

Server cons:
- we need to implement a clipping alternative for all present and future backends that show this behavior
- different behaviors for different backends 

Client pros:
- same code, same behavior for all platforms and backends
- it would be easier to implement other features, such as removing bad values, like SafetyNet does
- easy to switch off completely if one wants

Client cons:
- one could potentially put some nodes after volumeAmpControl. Examples:`Synth.tail(0, \default)`, or `Synth.after(s, \default)` before volumeAmpControl is created. Maybe that's not so easy to do after all?
- Recorder will put its node at tail(0), so after volumeAmpControl. This means that clipping will affect recording. But maybe there could be a "pre-fader" option for recorder, or anyway the behavior of "after default group" utilities should be figured out (Recorder, ServerMeter, Volume, any others?)

**Comment by dyfer** [comment]
I'd add one more Server-side pro / Client-side con: server-side implementation will also provide basic safety for other clients.

**Comment by elgiano** [comment]
I pushed a template-based implementation:
- clipping can be transparently disabled by setting `s.options.safetyClipThreshold = inf`
- implementation is through templates: the correct (clipping or not-clipping) function is selected at DriverStart time. No extra checks are performed after that moment
- code duplication is completely avoided

The cost is that now we have two audioIOProc funcs and two SC_CoreAudioDriver:Run funcs (from templates, no code duplication), which I don't think is a cost at all, or very minimal. Feedback would be welcome and very important :)

This doesn't deal with s.volume (which I would leave to the sclang-side stage limiter implementation) or other platforms than macOS. Since so far we don't have that many backends, would it be reasonable to double-check all of them and add this feature only to the ones that need it, and then evaluate every future backend addition (e.g. pulse)?

File: include/server/SC_WorldOptions.h [hunk 2]
```diff
@@ -49,6 +49,7 @@ struct WorldOptions {
 
     bool mRealTime = true;
     bool mMemoryLocking = false;
+    float mSafetyClipThreshold = 2.0;
```
Comment (brianlheim) [inline include/server/SC_WorldOptions.h:52]: note -- look for convo w @Spacechild1 about ABI/API changes

Comment (Spacechild1) [reply include/server/SC_WorldOptions.h:52]: thanks @brianlheim for catching that! `WorldOptions` is part of the public libscsynth API and changing a public struct causes API/ABI breakage.

Note that the libscsynth API/ABI has already been broken in past. For example, https://github.com/supercollider/supercollider/commit/695aea5b932d5915eeca629803d9565016ff2e2d changed function signatures, and https://github.com/supercollider/supercollider/commit/22c9deac2e38f6446bf254c55157925852c96142 turned `WorldOptions` from a old-style C++ struct to a C++11 struct. The first commit causes a linker error if header and binary don't have the same version. The second commit has no consequences expect possible compilation errors for pre-C++11 projects. 

However, changing the layout of a public struct can cause runtime crashes, e.g. when the host uses a new libsynth binary with the old header. This means it is now required that the `SC_WorldOptions.h` header and the libscsynth binary have the same SC version (which means you can't update the libscsynth binary without recompiling the host app against the same SuperCollider SDK version, and vice versa!) This is just something to be aware of.

I see two solutions:
a) make it clear that header and binary must have the same version
b) add a simple versioning mechanism similar to the SC plugin API, so that `World_New` fails on a version mismatch.

File: server/scsynth/SC_CoreAudio.cpp [hunk 3]
```diff
@@ -1598,6 +1606,9 @@ bool SC_CoreAudioDriver::DriverStart() {
     if (mWorld->mVerbosity >= 1) {
         scprintf("->SC_CoreAudioDriver::DriverStart\n");
     }
+
+    auto appIOProcFunc = mSafetyClipThreshold < INFINITY ? appIOProc<true> : appIOProc<false>;
```
Comment (brianlheim) [inline server/scsynth/SC_CoreAudio.cpp:1610]: i think you should also disable clipping if the threshold is negative. it would rhyme with the 'negative = off' behavior that pops up in other places in SC. and also, the clipping behavior will not work if the input is negative; if i'm following the definition of sc_clip2 right, the output will just sit at the negative value most of the time, which could lead to loud pops as well.

Comment (brianlheim) [inline server/scsynth/SC_CoreAudio.cpp:1610]: i think you should also disable clipping if the threshold is negative. it would rhyme with the 'negative = off' behavior that pops up in other places in SC. and also, the clipping behavior will not work if the input is negative; if i'm following the definition of sc_clip2 right, the output will just sit at the negative value most of the time, which could lead to loud pops as well.

File: server/scsynth/SC_CoreAudio.h [hunk 4]
```diff
@@ -268,8 +270,14 @@ class SC_CoreAudioDriver : public SC_AudioDriver {
 
     bool StopStart();
 
+    template <bool isClipping>
```
Comment (brianlheim) [inline server/scsynth/SC_CoreAudio.h:273]: typically we capitalize the first letter of template parameter names, to help distinguish them from ordinary variables

File: server/scsynth/SC_CoreAudio.h [hunk 5]
```diff
@@ -268,8 +270,14 @@ class SC_CoreAudioDriver : public SC_AudioDriver {
...
     void Run(const AudioBufferList* inInputData, AudioBufferList* outOutputData, int64 oscTime);
 
+    template <bool isClipping>
+    friend OSStatus appIOProc(AudioDeviceID device, const AudioTimeStamp* inNow, const AudioBufferList* inInputData,
+                              const AudioTimeStamp* inInputTime, AudioBufferList* outOutputData,
+                              const AudioTimeStamp* inOutputTime, void* defptr);
```
Comment (brianlheim) [inline server/scsynth/SC_CoreAudio.h:279]: why did you add this declaration to the header? it doesn't seem necessary, since appIOProc doesn't access any members in this class

File: server/scsynth/scsynth_main.cpp [hunk 6]
```diff
@@ -80,6 +80,7 @@ void Usage() {
              "          UDP ports never require passwords, so for security use TCP.\n"
              "   -N <cmd-filename> <input-filename> <output-filename> <sample-rate> <header-format> <sample-format>\n"
 #ifdef __APPLE__
+             "   -s <safety-clip-threshold>          (default %d)\n"
```
Comment (brianlheim) [inline server/scsynth/scsynth_main.cpp:83]: you should definitely add a note here that 'inf' (or negative number if you agree with my comment above) will disable it. and maybe add a brief explanation of what the option does here too.

File: include/server/SC_WorldOptions.h (hunk 2 above)
Comment (brianlheim) [reply include/server/SC_WorldOptions.h:52]: thanks @Spacechild1 !!

i don't know if many people use this part of the API at all. just thinking back, i have never seen or heard any concerns over the shared-libscsynth build breaking -- except for the Arch Linux distribution, but David Runge said he's not aware of anyone making use of it either. i've also never seen complaints about crashes or other breakage when those API changes did happen.

i wouldn't want to spend time on adding a feature that nobody will make use of, so for now i think we could just document in the header that the API may change across minor versions, so it should be compiled against headers from the same minor version used to link. that's at least a small stability guarantee. what do you think?

File: HelpSource/Classes/ServerOptions.schelp [hunk 7]
```diff
@@ -140,6 +140,11 @@ A Boolean indicating whether the server should try to lock its memory into physi
...
 An Integer indicating the maximum number of clients which can simultaneously receive notifications from the server. When using TCP this is also the maximum number of simultaneous connections. This is also used by the language to split ranges of link::Classes/Node##Nodes::, link::Classes/Buffer##Buffers::, or link::Classes/Bus##Busses::. In multi-client situations you will need to set this to at least the number of clients you wish to allow. This must be the same in the Server instances on every client. The default is 1.
 
+method:: safetyClipThreshold
+note:: MacOS only::
+A Float indicating a safety threshold for output values to be clipped. This is necessary on MacOSX because setting a low system volume doesn't prevent output values greater than +- 1 (which can happen by mistake, e.g. when sending a negative coefficient to a filter) to sound absolutely loud, since CoreAudio applies only a relative attenuation. Values are clipped only before being written to hardware output busses, so although this affects inter-app routing as well (since it would use those same busses), it doesn't affect recording, which still benefits from the full 32bit float range without any clipping.
```
Comment (brianlheim) [inline HelpSource/Classes/ServerOptions.schelp:145]: here are a few small suggestions:

```suggestion
A Float indicating a safety threshold for output values to be clipped to. This is necessary on macOS because setting a low system volume doesn't prevent output values greater than +/- 1 (which can happen by mistake, e.g. when sending a negative coefficient to a filter) to sound extremely loud, since CoreAudio applies only a relative attenuation. With this threshold, values are clipped just before being written to hardware output busses, so although this affects routing to other apps, it doesn't affect recording, which still benefits from the full 32-bit float range without any clipping.
```

**Review by brianlheim — CHANGES_REQUESTED** [review]
looks good overall, thanks!

**Comment by dyfer** [comment]
> Moreover, if the intention is ear safety, the end-stage clipper needs to be based on the server's volume setting, not full scale.
@jamshark70 I thought about this one and I don't really see it this way. If you use a DAW, and you turn town master fader, you don't expect to get signals only up to the level set by the fader, do you?

@elgiano I think the default headroom could be lower. @jamshark70 suggested 2 dB, (~`1.25`), I'd be fine with that, or maybe `1.5` (~3.5 dB), what do you think?

File: include/server/SC_WorldOptions.h (hunk 2 above)
Comment (Spacechild1) [reply include/server/SC_WorldOptions.h:52]: > i don't know if many people use this part of the API at all

I'm sure there are a few people who use libscsynth, but they probably all use a binary + header with the same version. So in practice, ABI breakage is not such a big deal.

> i've also never seen complaints about crashes or other breakage when those API changes did happen.

The difference is that previous breakages would lead to a compiler resp. linker error, which are easy to spot/fix. Changing the struct can cause runtime crashes, which are harder to debug.

> so for now i think we could just document in the header that the API may change across minor versions, so it should be compiled against headers from the same minor version used to link. that's at least a small stability guarantee. what do you think?

That would be option a), and I think that's probably sufficient :-)

File: server/scsynth/SC_CoreAudio.cpp (hunk 3 above)
Comment (elgiano) [reply server/scsynth/SC_CoreAudio.cpp:1610]: You're right, I think I got confused with dBs vs amp. Actually I think I intended to implement this interface in dBs, such as in `scsynth -s -6` or `scsynth -s 0` and then convert to amp before assignment. It feels more consistent with server.volume... but now I'm not so sure about it anymore.. any comment? Would it be better in amp or db for you?

File: server/scsynth/SC_CoreAudio.h (hunk 5 above)
Comment (elgiano) [reply server/scsynth/SC_CoreAudio.h:279]: You're right! I remember I had to do it at some point, but looking again at it I can't see why.. I should have definitely put a comment there. I will try to remove it and figure this out, thanks!

**Comment by elgiano** [comment]
Thanks @dyfer, I got confused with dBs vs amp. Please see https://github.com/supercollider/supercollider/pull/5110#discussion_r528309104 above :)

File: HelpSource/Classes/ServerOptions.schelp (hunk 7 above)
Comment (elgiano) [reply HelpSource/Classes/ServerOptions.schelp:145]: Thanks!!

File: server/scsynth/SC_CoreAudio.h (hunk 5 above)
Comment (elgiano) [reply server/scsynth/SC_CoreAudio.h:279]: Ok, I've figured it out. That friend function was already there, and it's needed (if I understand correctly) for appIOProc to access private members of `*defptr`, such as `mInputBufList` here:
https://github.com/supercollider/supercollider/blob/76e595cda18a3fd760a38dd574bc93af92880ba0/server/scsynth/SC_CoreAudio.cpp#L1264

the only thing this PR added to the declaration is a `template <bool IsClipping>` statement.

File: include/server/SC_WorldOptions.h (hunk 2 above)
Comment (brianlheim) [reply include/server/SC_WorldOptions.h:52]: @elgiano can you please add this note?

File: server/scsynth/SC_CoreAudio.h (hunk 5 above)
Comment (brianlheim) [reply server/scsynth/SC_CoreAudio.h:279]: you should add the `template` part to the preexisting friend declaration rather than creating a new one

File: server/scsynth/SC_CoreAudio.cpp [hunk 8]
```diff
@@ -1598,6 +1606,9 @@ bool SC_CoreAudioDriver::DriverStart() {
     if (mWorld->mVerbosity >= 1) {
         scprintf("->SC_CoreAudioDriver::DriverStart\n");
     }
+
+    auto appIOProcFunc = (mSafetyClipThreshold > 0 && mSafetyClipThreshold < INFINITY) ? appIOProc<true> : appIOProc<false>;
```
Comment (brianlheim) [inline server/scsynth/SC_CoreAudio.cpp:1610]: you've changed this one but not the other on line 1798. perhaps you should factor this into a member function?

File: server/scsynth/SC_CoreAudio.cpp (hunk 3 above)
Comment (brianlheim) [reply server/scsynth/SC_CoreAudio.cpp]: i don't have a strong opinion either way tbh, but i think working with amp over db is slightly easier for new users to understand.

Comment (dyfer) [reply server/scsynth/SC_CoreAudio.cpp]: I have a slight preference to keep this in amp, not in dB. 
I never used server.volume programmatically - it makes sense that it's in dB, but pretty much everything else is in amp, so to me it feels more like an outlier.

File: server/scsynth/SC_CoreAudio.h (hunk 5 above)
Comment (elgiano) [reply server/scsynth/SC_CoreAudio.h:279]: ops, now I see it! Thanks for catching it, my bad

**Comment by brianlheim** [comment]
@elgiano thanks for pushing changes! these two comments still need to be addressed, i realized they might be easy to miss since they happened in different reviews:

https://github.com/supercollider/supercollider/pull/5110/files#r528360213
https://github.com/supercollider/supercollider/pull/5110/files#r528359779

File: server/scsynth/SC_CoreAudio.h [hunk 9]
```diff
@@ -246,6 +249,8 @@ class SC_CoreAudioDriver : public SC_AudioDriver {
                                         AudioBufferList* outOutputData, const AudioTimeStamp* inOutputTime,
                                         void* defptr);
 
+    const bool isClippingEnabled() { return mSafetyClipThreshold > 0 && mSafetyClipThreshold < INFINITY; }
```
Comment (brianlheim) [inline server/scsynth/SC_CoreAudio.h:252]: `const` on the return type doesn't have any effect, but `const` on the function indicates that the function doesn't mutate the object; i think that's what you intended:

```suggestion
    bool isClippingEnabled() const { return mSafetyClipThreshold > 0 && mSafetyClipThreshold < INFINITY; }
```

File: include/server/SC_WorldOptions.h (hunk 2 above)
Comment (elgiano) [reply include/server/SC_WorldOptions.h]: Something like this?
```
// PLEASE NOTE:
// This API might change across minor versions.
// Always make sure, when using libscsynth, that the binary and this header come from the same minor version.
#

**Comment by elgiano** [comment]
Thanks @dyfer and @brianlheim. I put a short summary here since navigating reviews is starting to become difficult for me:
- I think I addressed everything: the only point left where I need comments is the warning in SC_WorldOptions.h. I put it in a review comment, now I can't find it anymore, so I'll copy it here. What do you think of:
```
// PLEASE NOTE:
// This API might change across minor versions.
// Always make sure, when using libscsynth, that the binary and this header come from the same SC minor version
```
- amp vs dB: I kept it in amp and changed the default threshold to 1.26 (ca. 2 dB)

**Comment by brianlheim** [comment]
> I put a short summary here since navigating reviews is starting to become difficult for me:

no worries, it is quite disordered. just one thing, it might help to collapse review comments with "mark as resolved" once you resolve each one, then it's clearer what still remains to be done.

> I think I addressed everything: the only point left where I need comments is the warning in SC_WorldOptions.h. I put it in a review comment, now I can't find it anymore, so I'll copy it here. What do you think of:

i think it's good. you should definitely also note that "this API" applies specifically to things in `server/include`, which i guess you could call the "libscsynth API". so that people don't get confused and think this also applies to the plugin API which has an actual versioning system. in fact, i would place this comment directly above the first line of code in all 3 files in `server/include`. also change "using libscsynth" to "using libscsynth as a shared library". libscsynth can also be built as a static lib (that's the default).

> SUPERNOVA: I added another switch to classlib's ServerOptions, to not pass the "-s" flag when using supernova, otherwise supernova would refuse to start. We will have to implement clipping for supernova as well, I hope it can be done quickly after we are clear with the scsynth implementation.

oh yes right i forgot about that! will you make the supernova implementation in another PR? it may be better for this change because there are already a lot of PR comments to sift through. i guess it will be done to the portaudio backend since there is no supernova coreaudio backend yet.

**Comment by elgiano** [comment]
> no worries, it is quite disordered. just one thing, it might help to collapse review comments with "mark as resolved" once you resolve each one, then it's clearer what still remains to be done.

This I usually leave to the one who made the first comment, it feels a bit rude for me to decide when it is resolved :)

> i think it's good. you should definitely also note that "this API" applies specifically to things in server/include, which i guess you could call the "libscsynth API". so that people don't get confused and think this also applies to the plugin API which has an actual versioning system. in fact, i would place this comment directly above the first line of code in all 3 files in server/include. also change "using libscsynth" to "using libscsynth as a shared library". libscsynth can also be built as a static lib (that's the default).

I've updated the message and added it to all three files in include/server

> oh yes right i forgot about that! will you make the supernova implementation in another PR? it may be better for this change because there are already a lot of PR comments to sift through. i guess it will be done to the portaudio backend since there is no supernova coreaudio backend yet.

I'll open a new PR as soon as we are done with this one

**Comment by dyfer** [comment]
just a note for supernova: we use portaudio on macOS so this is the backend for which this needs to be implemented. It would be good to test whether windows APIs have the same problem (namely MME as this is the default) and then decide if we apply it only for mac or mac and windows.

**Comment by brianlheim** [comment]
> This I usually leave to the one who made the first comment, it feels a bit rude for me to decide when it is resolved :)

i'm used to the opposite workflow but i don't mind doing it that way, thanks for explaining!

**Review by brianlheim — APPROVED** [review]
thanks!

**Comment by elgiano** [comment]
Sorry, I didn't expect to have linting problems with comments, but ofc it makes sense! it's clang-formatted now :)

**Comment by elgiano** [comment]
@timblechmann would it be possible for you to get in touch about implementing this for supernova? I'd really benefit from a quick chat about it!

**Review by brianlheim — APPROVED** [review]
thanks! gonna let someone else approve/merge as this is a slightly complex feature.

**Comment by brianlheim** [comment]
fwiw @elgiano your branch looks good to me aside from the capitalization missing here https://github.com/elgiano/supercollider/commit/aba9f872d4152741a7c2b667c11b07b2f3c57278#diff-95e496ab2ce1c36e8c20a690148fa7def521da35f7d1fac6e3cd2917241609e3R290

File: HelpSource/Classes/ServerOptions.schelp [hunk 10]
```diff
@@ -142,6 +142,11 @@ A Boolean indicating whether the server should try to lock its memory into physi
...
 
+method:: safetyClipThreshold
+note:: MacOS only::
+A Float indicating a safety threshold for output values to be clipped to. This is necessary on macOS because setting a low system volume doesn't prevent output values greater than +/- 1 (which can happen by mistake, e.g. when sending a negative coefficient to a filter) to sound extremely loud, since CoreAudio applies only a relative attenuation. With this threshold, values are clipped just before being written to hardware output busses, so although this affects routing to other apps, it doesn't affect recording, which still benefits from the full 32-bit float range without any clipping.
+
+Defaults to a threshold of 1.26 (ca. 2 dB), to save some ears and still allow some headroom. Setting safetyClipThreshold to strong::inf:: disables clipping altogether.
```
Comment (dyfer) [inline HelpSource/Classes/ServerOptions.schelp:149]: I tested this and it works as desired, thank you so much @elgiano and @brianlheim ! This is a HUGE improvement!

Before merging I wanted to propose a small change to the documentation:

> This is necessary on macOS because setting a low system volume doesn't prevent output values greater than +/- 1 (which can happen by mistake, e.g. when sending a negative coefficient to a filter) to sound extremely loud, since CoreAudio applies only a relative attenuation. 

to sound extremely loud -> from sounding extremely loud

and maybe changing the order and dropping the part about CoreAudio:

>This is necessary on macOS because setting a low system volume doesn't prevent output values greater than +/- 1 from sounding extremely loud, which can happen by mistake, e.g. when sending a negative coefficient to a filter. 

Regarding this:

> With this threshold, values are clipped just before being written to hardware output busses, so although this affects routing to other apps, it doesn't affect recording, which still benefits from the full 32-bit float range without any clipping.

Routing sound between apps on macOS is more of an exception than a rule. And also server recording might or might not be set to 32-bit float. How about rephrasing this as:

> With this threshold, values are clipped just before being written to hardware output busses, which does not affect the recording. However, the signal will be affected if it's above the threshold and  the sound is routed to other apps using 3rd-party software.

And finally:

> Defaults to a threshold of 1.26 (ca. 2 dB), to save some ears and still allow some headroom. Setting safetyClipThreshold to strong::inf:: disables clipping altogether.

I think it's both `inf` and `0`, right? I'd list both here. I'd also use `code:: ::` instead of `strong:: ::` for it.

**Review by dyfer — COMMENTED** [review]
I'm happy to merge it as is, but if you feel like making these changes I think it would be even better. Thanks!

File: HelpSource/Classes/ServerOptions.schelp (hunk 10 above)
Comment (elgiano) [reply HelpSource/Classes/ServerOptions.schelp]: Thanks @dyfer, your changes make it look clearer to me, I just included them all

**Comment by elgiano** [comment]
@brianlheim 
> fwiw @elgiano your branch looks good to me aside from the capitalization missing here elgiano@aba9f87#diff-95e496ab2ce1c36e8c20a690148fa7def521da35f7d1fac6e3cd2917241609e3R290

Thanks for looking at this Brian, and catching that typo! I haven't tested it yet (haven't built it once actually) because at this moment my access to a mac computer is somehow limited. But I will get to do it soon!

**Review by dyfer — APPROVED** [review]
I think this looks good, thank you!

IMO can be merged once CI passes (or maybe earlier? since last change did not touch cpp/sc code)

**Comment by elgiano** [comment]
Thanks everyone, I'm very happy this got merged :+1:

**Comment by dyfer** [comment]
@elgiano are you still planning to make the supernova PR for this functionality? IMO we should have this on our radar for 3.12

**Comment by elgiano** [comment]
I had a branch going on for supernova, couldn't really get it to work though, and then I ended up not having a Mac anymore.... however I might get one from my job soon and then I would look again into it!
If anyone wants to look at my branch, here it is: https://github.com/elgiano/supercollider/tree/topic/macos-supernova-clip

**Comment by joshpar** [comment]
@elgiano - how is your progress on this looking? More than happy to take it over if you aren't able to keep on it.

**Comment by elgiano** [comment]
@joshpar 
Thanks! I would be super happy if you could have a look at it! You can see all the changes here:
https://github.com/supercollider/supercollider/compare/develop...elgiano:topic/macos-supernova-clip

I still don't have a working Mac right now, so I can't do anything about it.
